azure/durable_functions/models/NewTask.py

```diff
@@ -0,0 +1,314 @@
+    instantiated on its own.
+    """
+
+    def __init__(self, id_: int, actions: Union[List[Action], Action]):
+        """Initialize the TaskBase.
+
+        Parameters
+        ----------
+        id_ : int
+            An ID for the task
+        actions : List[Any]
+            The list of DF actions representing this Task.
+            Needed for reconstruction in the extension.
+        """
+        self.id: int = id_
+        self.state = TaskState.RUNNING
+        self.was_yielded: bool = False
+        self.parent: Optional[CompoundTask] = None
+
+        self.value: Any = None
+        self.action_repr: Union[List[Action], Action] = actions
+        self.is_played = False
+
+    def set_is_played(self, is_played: bool):
+        """Set the is_played flag for the Task.
+
+        Needed for updating the orchestrator's is_replaying flag.
+
+        Parameters
+        ----------
+        is_played : bool
+            Whether the latest event for this Task has been played before.
+        """
+        self.is_played = is_played
+
+    def change_state(self, state: TaskState):
+        """Transition a running Task to a terminal state: success or failure.
+
+        Parameters
+        ----------
+        state : TaskState
+            The terminal state to assign to this Task
+
+        Raises
+        ------
+        Exception
+            When the input state is RUNNING
+        """
+        if state is TaskState.RUNNING:
+            raise Exception("Cannot change Task to the RUNNING state.")
+        self.state = state
+
+    def set_value(self, is_error: bool, value: Any):
+        """Set the value of this Task: either an exception of a result.
+
+        Parameters
+        ----------
+        is_error : bool
+            Whether the value represents an exception of a result.
+        value : Any
+            The value of this Task
+
+        Raises
+        ------
+        Exception
+            When the Task failed but its value was not an Exception
+        """
+        new_state = self.state
+        if is_error:
+            if not isinstance(value, Exception):
+                raise Exception(f"Task ID {self.id} failed but it's value was not an Exception")
+            new_state = TaskState.FAILED
+        else:
+            new_state = TaskState.SUCCEEDED
+        self.change_state(new_state)
+        self.value = value
+        self.propagate()
+
+    def propagate(self):
+        """Notify parent Task of this Task's state change."""
+        has_completed = not (self.state is TaskState.RUNNING)
+        has_parent = not (self.parent is None)
+        if has_completed and has_parent:
+            self.parent.handle_completion(self)
+
+
+class CompoundTask(TaskBase):
+    """A Task of Tasks.
+
+    Contains shared logic that drives all of its sub-classes.
+    Should never be instantiated on its own.
+    """
+
+    def __init__(self, tasks: List[TaskBase], action_wrapper=None):
```

Comment:
Yeah I need to improve the naming here, but you got the gist of it right.
An `action_wrapper` is either a `WhenAnyAction` or a `WhenAllAction` constructor.
You could say that it `wraps` around a list of tasks because it "contains them within".

I realize this is awkward design, but I'm slightly forced to do this in order to support the V1 and V2 versions of the OOProc replay protocol. You've probably also realized that every time we have to modify an actions array, be it either inside the orchestration-context object or inside a task, the logic is somewhat complicated.

I think there's a clean way to address this problem though. I think I'll force the SDK-internal representation of actions to always use the V2 approach and then, right before serializing to JSON, I'll decide whether to translate the representation to V1. I think that will keep the backwards-compatibility logic in fewer places. What do you think?

Reply:
Let's focus on that in a subsequent PR. I like the idea of that approach, but it may be expensive, so we may want to wait until extension bundles are using V2 schema...